client/contentrouting.go

```diff
@@ -0,0 +1,54 @@
+package client
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+type ContentRoutingClient struct {
+	client DelegatedRoutingClient
+}
+
+func NewContentRoutingClient(c DelegatedRoutingClient) *ContentRoutingClient {
+	return &ContentRoutingClient{client: c}
+}
+
+func (c *ContentRoutingClient) Provide(context.Context, cid.Cid, bool) error {
+	return fmt.Errorf("content providing is not supported by this version of delegated routing")
+}
+
+func (c *ContentRoutingClient) FindProvidersAsync(ctx context.Context, key cid.Cid, numResults int) <-chan peer.AddrInfo {
+	addrInfoCh := make(chan peer.AddrInfo)
+	resultCh, err := c.client.FindProvidersAsync(ctx, key)
+	if err != nil {
+		close(addrInfoCh)
+		return addrInfoCh
+	}
+	go func() {
```

Comment:
Thoughts (not blockers for merge):

Not sure if this is avoidable but wanted your thoughts on this. It looks like every abstraction layer involving a channel turns into another goroutine to pull `x` from the channel and return `f(x)` into a new channel.

Maybe once generics become more widely supported in our codebase (once Go 1.19 is released) this will become easier by letting us return generic channels and pass in functions to convert between types so a goroutine isn't required.

To be fair short lived goroutines aren't necessarily a big issue unless we're doing them on mass so we can cross this bridge when it becomes a problem.

Reply:
Yes. Without generics, we would need to resort to empty interfaces, which I think makes the code very unmanageable. I lean towards code readability unless there is a clear performance problem. And indeed it seems like Go generics can fix this issue.